exchange/utils.go

```diff
@@ -741,3 +765,41 @@ func applyFPD(fpd *firstpartydata.ResolvedFirstPartyData, bidReq *openrtb2.BidRe
 		bidReq.User = fpd.User
 	}
 }
+
+func buildBidderRequestsWithBidResponses(req AuctionRequest, aliases map[string]string) map[openrtb_ext.BidderName]BidderRequest {
+	var bidderToBidderResponse map[openrtb_ext.BidderName]BidderRequest
+	if len(req.StoredBidResponses) > 0 {
+		// delete imps with stored bid resp
+		imps := req.BidRequest.Imp
+		req.BidRequest.Imp = nil //or new slice?
+		for _, imp := range imps {
+			if _, ok := req.StoredBidResponses[imp.ID]; !ok {
+				//add real imp back to request
+				req.BidRequest.Imp = append(req.BidRequest.Imp, imp)
+			}
+		}
+		// bidder -> imp id -> stored bid resp
+		bidderToBidderResponse = make(map[openrtb_ext.BidderName]BidderRequest)
+		for impID, storedData := range req.StoredBidResponses {
+			for bidderName, storedResp := range storedData {
+				if _, ok := bidderToBidderResponse[openrtb_ext.BidderName(bidderName)]; !ok {
+					//new bidder with stored bid responses
+					//!!! check if bidder is valid/exists
```

Comment:
Is this comment still needed, or if it is, could those `!!!` be removed?

Reply:
Removed and added this validation to 'processStoredAuctionResponses'